src/python/dxpy/sugar/transfers.py

```diff
@@ -0,0 +1,952 @@
+
+    if remote_filename:
+        kwargs["name"] = remote_filename
+    if not remote_folder:
+        remote_folder = "/"
+    elif not remote_folder.startswith("/"):
+        remote_folder = "/{}".format(remote_folder)
+    kwargs["folder"] = remote_folder
+    if project:
+        kwargs["project"] = project
+    LOG.info("Uploading file %s to %s", filename, remote_filename or filename)
+    handler = dxpy.upload_local_file(filename, **kwargs)
+    if return_handler:
+        return handler
+    else:
+        return dxpy.dxlink(handler.get_id(), handler.describe()["project"])
+
+
+def compress_and_upload_file(
+    filename,
+    remote_filename=None,
+    remote_folder="/",
+    compression_level=1,
+    return_handler=False,
+    project=None,
+    wait_on_close=False,
+    method="gz",
+    max_part_size=None
+):
+    """
+    Gzip and upload a local file.
+
+    Shorthand for running 'gzip' and 'dx upload' using a subprocess on a given
+    local file.
+
+    Args:
+        filename (str): Local filename.
+        remote_filename (str): Optional, remote filename to upload file to.
+        remote_folder (str): Optional, remote folder to upload file to.
+        compression_level (int): Level of compression between 1 and 9 to compress
+            file to. Specify 1 for gzip --fast and 9 for gzip --best. If not
+            specified, --fast is assumed.
+        return_handler (bool): Whether to return a DXFile handler.
+        project (str): The project ID to upload to, if not the currently selected
+            project.
+        wait_on_close (bool): Whether to block until the file has closed.
+        method (str): Compression method; one of 'gz', 'bz2'.
+        max_part_size (int): Optional, maximum file part size, defaults to project
+            value.
+
+    Returns:
+        DNAnexus link or DXFile pointing to the uploaded file (depending on
+        the value of `return_handler`).
+
+    Raises:
+        ValueError: if compression_level not between 1 and 9
+        CalledProcessError: propogated from run_pipe if called command fails
+    """
+    ext = ".{}".format(method)
+    if remote_filename is None:
+        remote_filename = os.path.basename(filename)
+    if not remote_filename.endswith(ext):
+        remote_filename += ext
+
+    if not remote_folder:
+        remote_folder = "/"
+    elif not remote_folder.startswith("/"):
+        remote_folder = "/{}".format(remote_folder)
+
+    remote_path = "{}/{}".format(remote_folder, remote_filename)
+    if project:
+        remote_path = "{}:{}".format(project, remote_path)
+
+    # check that compression level is valid
+    if compression_level < 1 or compression_level > 9:
+        raise ValueError("Compression level must be between 1 and 9")
+
+    if not max_part_size or max_part_size <= 0:
+        max_part_size = _get_max_part_size(project)
+
+    if filename.endswith(ext):
+        cmd = [
+            "dx", "upload", "--brief", "--buffer-size", str(max_part_size),
```

Comment:
Not sure if I mentioned this in the previous PR.

Underneath, we try and perform uploads in parallel with each thread using the `buffer-size` that was set (code [permalink](https://github.com/dnanexus/dx-toolkit/blob/a86d952d5142dc01a18d64f50d5049a0f7e39dbe/src/python/dxpy/bindings/dxfile.py#L513)). We submit these to task to worker pools. Which AFAIK is unaware of the current memory on the worker. What can happen is:

```
num_active_threads * buffer-size > memory availble on machine
```

While your change is less prone failing to upload in `partLimits` I think you're more prone to making the worker going OOM.

My initial idea for a fix is:
1. Make the initial `buffer-size` calculation _slightly_ worker aware (i.e never choose a buffer size greater than the memory on the worker).
2. Modify writes to be aware of the pool and current memory on the machine.

Maybe there's a simpler fix.

Reply:
Cool. I used psutil to get the available memory, and have Uploader._submit pass in its _max_workers attribute to the upload function, and then use those together to limit max_part_size to at most available_memory / max_workers.